src/Avalonia.Controls/Border.cs

```diff
@@ -120,29 +110,180 @@ protected override Size ArrangeOverride(Size finalSize)
         {
             if (Child != null)
             {
-                var padding = Padding + new Thickness(BorderThickness);
+                var padding = Padding + BorderThickness;
                 Child.Arrange(new Rect(finalSize).Deflate(padding));
             }
 
+            _borderRenderer.Update(finalSize, BorderThickness, CornerRadius);
+
             return finalSize;
         }
 
         internal static Size MeasureOverrideImpl(
             Size availableSize,
             IControl child,
             Thickness padding,
-            double borderThickness)
+            Thickness borderThickness)
         {
-            padding += new Thickness(borderThickness);
+            padding += borderThickness;
 
             if (child != null)
             {
                 child.Measure(availableSize.Deflate(padding));
                 return child.DesiredSize.Inflate(padding);
             }
-            else
+
+            return new Size(padding.Left + padding.Right, padding.Bottom + padding.Top);
+        }
+
+        internal class BorderRenderer
```

Comment:
I think I'd rather have this class moved out of `Border`. Maybe put it as an internal class under `Utils`?

Reply:
Will have a look at present render tests and create new ones to represent my additions. That way we can make sure everything looks as it should be.